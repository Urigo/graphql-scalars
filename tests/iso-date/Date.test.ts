/**
 * Copyright (c) 2017, Dirk-Jan Rutten
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import GraphQLDate from '../../src/resolvers/iso-date/Date'
import { Kind } from 'graphql'
// flowlint-next-line untyped-import:off
import { stringify } from 'jest-matcher-utils'

const invalidDates = [
  'invalid date',
  '2015-02-29'
]

const validDates = [
  [ '2016-12-17', new Date(Date.UTC(2016, 11, 17)) ],
  [ '2016-02-01', new Date(Date.UTC(2016, 1, 1)) ]
]

describe('GraphQLDate', () => {
  it('has a description', () => {
    expect(GraphQLDate.description).toMatchSnapshot()
  })

  describe('serialization', () => {
    [
      {},
      [],
      null,
      undefined,
      true
    ].forEach(invalidInput => {
      it(`throws error when serializing ${stringify(invalidInput)}`, () => {
        expect(() =>
          GraphQLDate.serialize(invalidInput)
        ).toThrowErrorMatchingSnapshot()
      })
    });

    [
      [ new Date(Date.UTC(2016, 11, 17, 14)), '2016-12-17' ],
      [ new Date(Date.UTC(2016, 0, 1, 14, 48, 10, 3)), '2016-01-01' ],
      [ new Date(Date.UTC(2016, 0, 1)), '2016-01-01' ]
    ].forEach(([ value, expected ]) => {
      it(`serializes javascript Date ${stringify(value)} into ${stringify(expected)}`, () => {
        expect(
          GraphQLDate.serialize(value)
        ).toEqual(expected)
      })
    })

    it(`throws error when serializing invalid javascript Date`, () => {
      expect(() =>
        GraphQLDate.serialize(new Date('invalid date'))
      ).toThrowErrorMatchingSnapshot()
    })

    // Serializes from date string
    validDates.forEach(([value]) => {
      it(`serializes date-string ${value}`, () => {
        expect(
          GraphQLDate.serialize(value)
        ).toEqual(value)
      })
    })

    invalidDates.forEach(dateString => {
      it(`throws an error when serializing an invalid date-string ${stringify(dateString)}`, () => {
        expect(() =>
          GraphQLDate.serialize(dateString)
        ).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe('value parsing', () => {
    validDates.forEach(([ value, expected ]) => {
      it(`parses date-string ${stringify(value)} into javascript Date ${stringify(expected)}`, () => {
        expect(
          GraphQLDate.parseValue(value)
        ).toEqual(expected)
      })
    });

    [
      4566,
      {},
      [],
      true,
      null
    ].forEach(invalidInput => {
      it(`throws an error when parsing ${stringify(invalidInput)}`, () => {
        expect(() =>
          GraphQLDate.parseValue(invalidInput)
        ).toThrowErrorMatchingSnapshot()
      })
    })

    invalidDates.forEach(dateString => {
      it(`throws an error parsing an invalid datetime-string ${stringify(dateString)}`, () => {
        expect(() =>
          GraphQLDate.parseValue(dateString)
        ).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe('literial parsing', () => {
    validDates.forEach(([ value, expected ]) => {
      const literal = {
        kind: Kind.STRING, value: value.toString(),
      }

      it(`parses literal ${stringify(literal)} into javascript Date ${stringify(expected)}`, () => {
        expect(
          GraphQLDate.parseLiteral(literal, {})
        ).toEqual(expected)
      })
    })

    invalidDates.forEach(value => {
      const invalidLiteral = {
        kind: Kind.STRING, value
      }
      it(`errors when parsing invalid literal ${stringify(invalidLiteral)}`, () => {
        expect(() =>
          GraphQLDate.parseLiteral(invalidLiteral, {})
        ).toThrowErrorMatchingSnapshot()
      })
    });

    [
      {
        kind: Kind.FLOAT, value: '5'
      },
      ({
        kind: Kind.DOCUMENT
      } as any)
    ].forEach(literal => {
      it(`errors when parsing invalid literal ${stringify(literal)}`, () => {
        expect(() =>
          GraphQLDate.parseLiteral(literal, {})
        ).toThrowErrorMatchingSnapshot()
      })
    })
  })
})
