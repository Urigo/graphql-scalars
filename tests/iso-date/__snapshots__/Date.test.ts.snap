// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLDate has a description 1`] = `"A date string, such as 2007-12-03, compliant with the \`full-date\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar."`;

exports[`GraphQLDate literial parsing errors when parsing invalid literal {"kind": "Document"} 1`] = `"Date cannot represent non string type false"`;

exports[`GraphQLDate literial parsing errors when parsing invalid literal {"kind": "FloatValue", "value": "5"} 1`] = `"Date cannot represent non string type 5"`;

exports[`GraphQLDate literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "2015-02-29"} 1`] = `"Date cannot represent an invalid date-string 2015-02-29."`;

exports[`GraphQLDate literial parsing errors when parsing invalid literal {"kind": "StringValue", "value": "invalid date"} 1`] = `"Date cannot represent an invalid date-string invalid date."`;

exports[`GraphQLDate serialization throws an error when serializing an invalid date-string "2015-02-29" 1`] = `"Date cannot represent an invalid date-string 2015-02-29."`;

exports[`GraphQLDate serialization throws an error when serializing an invalid date-string "invalid date" 1`] = `"Date cannot represent an invalid date-string invalid date."`;

exports[`GraphQLDate serialization throws error when serializing [] 1`] = `"Date cannot represent a non string, or non Date type []"`;

exports[`GraphQLDate serialization throws error when serializing {} 1`] = `"Date cannot represent a non string, or non Date type {}"`;

exports[`GraphQLDate serialization throws error when serializing invalid javascript Date 1`] = `"Date cannot represent an invalid Date instance"`;

exports[`GraphQLDate serialization throws error when serializing null 1`] = `"Date cannot represent a non string, or non Date type null"`;

exports[`GraphQLDate serialization throws error when serializing true 1`] = `"Date cannot represent a non string, or non Date type true"`;

exports[`GraphQLDate serialization throws error when serializing undefined 1`] = `"Date cannot represent a non string, or non Date type undefined"`;

exports[`GraphQLDate value parsing throws an error parsing an invalid datetime-string "2015-02-29" 1`] = `"Date cannot represent an invalid date-string 2015-02-29."`;

exports[`GraphQLDate value parsing throws an error parsing an invalid datetime-string "invalid date" 1`] = `"Date cannot represent an invalid date-string invalid date."`;

exports[`GraphQLDate value parsing throws an error when parsing [] 1`] = `"Date cannot represent non string type []"`;

exports[`GraphQLDate value parsing throws an error when parsing {} 1`] = `"Date cannot represent non string type {}"`;

exports[`GraphQLDate value parsing throws an error when parsing 4566 1`] = `"Date cannot represent non string type 4566"`;

exports[`GraphQLDate value parsing throws an error when parsing null 1`] = `"Date cannot represent non string type null"`;

exports[`GraphQLDate value parsing throws an error when parsing true 1`] = `"Date cannot represent non string type true"`;
